sources/api/shibaken/src/main.rs

```diff
@@ -45,75 +38,56 @@ impl UserData {
-    let uri = IMDS_SESSION_TOKEN_URI;
-    let imds_session_token = client
-        .put(uri)
-        .header("X-aws-ec2-metadata-token-ttl-seconds", "60")
-        .send()
-        .context(error::ImdsRequest { method: "PUT", uri })?
-        .error_for_status()
-        .context(error::ImdsResponse { uri })?
-        .text()
-        .context(error::ImdsText { uri })?;
-    Ok(imds_session_token)
-}
-
-/// Helper to fetch data from IMDS. Inspired by pluto.
-fn fetch_from_imds(client: &Client, uri: &str, session_token: &str) -> Result<Option<String>> {
-    let response = client
-        .get(uri)
-        .header("X-aws-ec2-metadata-token", session_token)
-        .send()
-        .context(error::ImdsRequest { method: "GET", uri })?;
-    if response.status().as_u16() == 404 {
-        return Ok(None);
-    }
-    Ok(Some(
-        response
-            .error_for_status()
-            .context(error::ImdsResponse { uri })?
-            .text()
-            .context(error::ImdsText { uri })?,
-    ))
-}
-
 /// Returns a list of public keys.
 fn fetch_public_keys_from_imds() -> Result<Vec<String>> {
-    info!("Fetching IMDS session token");
-    let client = Client::new();
-    let imds_session_token = fetch_imds_session_token(&client)?;
-
+    info!("Connecting to IMDS");
+    let mut imdsclient = ImdsClient::new().context(error::ImdsClient)?;
     info!("Fetching list of available public keys from IMDS");
     // Returns a list of available public keys as '0=my-public-key'
-    let public_key_list = if let Some(public_key_list) =
-        fetch_from_imds(&client, IMDS_PUBLIC_KEY_BASE_URI, &imds_session_token)?
-    {
-        debug!("available public keys '{}'", &public_key_list);
-        public_key_list
-    } else {
-        debug!("no available public keys");
-        return Ok(Vec::new());
+    let public_key_list = match imdsclient.fetch_metadata("public-keys", "public keys") {
+        Ok(Some(public_key_list)) => {
+            debug!("available public keys '{}'", &public_key_list);
+            public_key_list
+        }
+        Ok(None) => {
+            debug!("no available public keys");
+            return Ok(Vec::new());
+        }
+        Err(e) => Err(e).context(error::ImdsRequest)?,
     };
 
-    info!("Generating uris to fetch text of available public keys");
-    let public_key_uris = build_public_key_uris(&public_key_list);
+    info!("Generating targets to fetch text of available public keys");
+    let public_key_targets = build_public_key_targets(&public_key_list);
 
     info!("Fetching public keys from IMDS");
     let mut public_keys = Vec::new();
-    for uri in public_key_uris {
-        let public_key_text = fetch_from_imds(&client, &uri, &imds_session_token)?
-            .context(error::KeyNotFound { uri })?;
-        let public_key = public_key_text.trim_end();
+    let target_count: u32 = 0;
+    for target in &public_key_targets {
+        let target_count = target_count + 1;
```

Comment:
But perhaps I'm misunderstanding the purpose here.

Reply:
This is just for some extra status context, taking advantage of the description variable that wasn't in the original `shibaken` IMDS logic, but is present in `early-boot-config`. This count is unrelated to the index number (for example if index `0` is the only available key then it will be `public key (1/1)`)

The generated log messages that look something like:
`Requesting public key (1/2) from http://169.254.169.254/latest/meta-data/public-keys/0/openssh-key`